oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentStore.java

```diff
@@ -1470,8 +1487,18 @@ public <T extends Document> boolean create(Collection<T> collection, List<Update
                 }
                 insertSuccess = true;
                 return true;
-            } catch (MongoException e) {
-                return false;
+            } catch (BsonMaximumSizeExceededException e) {
+                T doct = null;
+                for (T doc : docs) {
+                    doct = doc;
+                    // doc.getMemory()/2 - converting from UTF-16 to UTF-8
+                    if (doc.getMemory()/2 > SIZE_LIMIT) {
```

Comment:
(I haven't followed all discussion) is UTF-16 always the case, or is there something that this could be made dependent upon ?

Reply:
Mongo returns it in UTF-16 so I have changed it to UTF-8

Reply:
I think the problem here is that we'd really like to estimate the memory usage within Mongo (which is UTF-8 IIUC), but the method used here is estimating the memory usage within the JVM (which does a multiply by 2, so would estimate for UTF-16). Is that correct?

The problem with the above is that it can potentially be off significantly - as the method used is optimized to estimate JVM side memory use, not Mongo side. And dividing by 2 could skew things..

I wonder if we should indeed lower the threshold for issuing a log.error here (and then phrase the message that the exact size might differ)..

Reply:
It was done to mention the document id that has caused the exception

Reply:
I think it's anyway difficult to convert our JVM estimate to a MongoDB estimate, for that we'd have to know how much space integers, boolean etc are taking up. With the current "divide by 2" we assume they also take half the space, which might or might not be true.

So I would add some margin to this condition, either by already warning at say 14MB - or, more drastically, by dividing by 3 (which would correspond to warning at 10.7MB already). That would be a bit more "loudly", as it would list IDs of more documents than necessary, but it would more likely not miss any. Without any margin there is chance to miss though.

Reply:
I need to add Warn if a doc size reaches 14MB as it won't though any exception. We can have a separate jira to add warn message when a document size reaches 14MB. WDYT?

Reply:
The 14MB was just to be on the very safe side, I wasn't suggesting to introduce an additional 14MB check.

Let's say you create a document with 16MB+1byte. Mongo will throw a BsonMaximumSizeExceededException. But for some reason our estimate of that document, divided by 2 is smaller than 16MB (our `getMemory` is only approximate). That case we wouldn't catch atm. But we could catch it as well by eg dividing by 3 or use a lower threshold (whatever is easier). (yes, we would potentially warn about documents that were not the reason mongo issued the exception, but I don't think that's a problem)